From c7d2c850cd4708eba02850906c5ef835859ab15e Mon Sep 17 00:00:00 2001
From: Mythri P K <mythri.p.k@intel.com>
Date: Fri, 25 Apr 2014 01:54:53 +0530
Subject: [PATCH 23/39] ASoC: soc-fw: add volsw_ext type for mixer

Issue: GMINL-3679

Volsw_ext type needs to be added to support any
control which has integer info type to be handled
for ex: volume and ramp duration.

Change-Id: I9dc83650b7a2d69b1268f8c1babeabd813f7b4bf
Signed-off-by: Mythri P K <mythri.p.k@intel.com>
---
 include/uapi/sound/asoc.h | 6 ++++++
 sound/soc/soc-fw.c        | 3 +++
 2 files changed, 9 insertions(+)

diff --git a/include/uapi/sound/asoc.h b/include/uapi/sound/asoc.h
index bc3891a..01f1fd5 100644
--- a/include/uapi/sound/asoc.h
+++ b/include/uapi/sound/asoc.h
@@ -330,6 +330,8 @@
 #define SOC_CONTROL_TYPE_STROBE		11
 #define SOC_CONTROL_TYPE_BYTES_EXT	12
 
+#define SOC_CONTROL_TYPE_VOLSW_EXT      14
+
 /* compound control IDs */
 #define SOC_CONTROL_IO_VOLSW \
 	SOC_CONTROL_ID(SOC_CONTROL_TYPE_VOLSW, \
@@ -379,6 +381,10 @@
 	SOC_CONTROL_ID(SOC_CONTROL_TYPE_EXT, \
 		SOC_CONTROL_TYPE_EXT, \
 		SOC_CONTROL_TYPE_BYTES_EXT)
+#define SOC_CONTROL_IO_VOLSW_EXT \
+	SOC_CONTROL_ID(SOC_CONTROL_TYPE_EXT, \
+		SOC_CONTROL_TYPE_EXT, \
+		SOC_CONTROL_TYPE_VOLSW)
 /* widget has no PM register bit */
 #define SND_SOC_NOPM	-1
 
diff --git a/sound/soc/soc-fw.c b/sound/soc/soc-fw.c
index 72f6b78..a934758 100644
--- a/sound/soc/soc-fw.c
+++ b/sound/soc/soc-fw.c
@@ -105,6 +105,8 @@ static const struct snd_soc_fw_kcontrol_ops io_ops[] = {
 	{SOC_CONTROL_IO_STROBE, snd_soc_get_strobe,
 		snd_soc_put_strobe, NULL},
 
+	{SOC_CONTROL_IO_VOLSW_EXT, NULL,
+		NULL, snd_soc_info_volsw},
 	{SOC_DAPM_IO_VOLSW, snd_soc_dapm_get_volsw,
 		snd_soc_dapm_put_volsw, NULL},
 	{SOC_DAPM_IO_ENUM_DOUBLE, snd_soc_dapm_get_enum_double,
@@ -813,6 +815,7 @@ static int soc_fw_kcontrol_load(struct soc_fw *sfw, struct snd_soc_fw_hdr *hdr)
 
 		switch (SOC_CONTROL_GET_ID_INFO(control_hdr->index)) {
 		case SOC_CONTROL_TYPE_VOLSW:
+		case SOC_CONTROL_TYPE_VOLSW_EXT:
 		case SOC_CONTROL_TYPE_STROBE:
 		case SOC_CONTROL_TYPE_VOLSW_SX:
 		case SOC_CONTROL_TYPE_VOLSW_S8:
-- 
1.9.1

